src/cms/views/pages/page_actions.py

```diff
@@ -112,6 +116,86 @@ def delete_page(request, page_id, region_slug, language_code):
     )
 
 
+# pylint: disable=unused-argument
+def link_callback(uri, rel):
+    """
+    according to xhtml2pdf documentation this function is neccessary for
+    resolving the django static files references and returning absolute paths
+    to the files
+
+    :param uri: uri that is generated by django template tag 'static'
+    :type uri: string
+    :return: the absolute path according to django's static file settings
+    """
+    sUrl = settings.STATIC_URL
+    sRoot = settings.STATIC_ROOT
+    mUrl = settings.MEDIA_URL
+    mRoot = settings.MEDIA_ROOT
+    if uri.startswith(mUrl):
+        path = os.path.join(mRoot, uri.replace(mUrl, ""))
+    elif uri.startswith(sUrl):
+        path = os.path.join(sRoot, uri.replace(sUrl, ""))
+    else:
+        return uri
+    # make sure that file exists
+    if not os.path.isfile(path):
+        raise Exception(
+            "media URI must start with %s or %s, but received uri was %s"
+            % (sUrl, mUrl, uri)
+        )
+    return path
+
+
+@login_required
+@region_permission_required
+# pylint: disable=unused-argument
+def export_pdf(request, region_slug, page_id, language_code):
+    """
+    function for handling a pdf export request for a single page
+    if the page is a root page the corresponding django template
+    inserts the content of all children pages
+
+    :param request: request submitted for rendering pdf document
+    :type request: HttpRequest
+    :param page_id: database id for selected page
+    :type page_id: integer
+    :param language_code: bcp47 code of the current language
+    :type language_code: string
+    :raises PermissionDenied: user login and permissions required
+    :return: pdf document offered for download
+    :rtype: HttpResponse
+    """
+    region = Region.get_current_region(request)
+    page = get_object_or_404(region.pages, pk=page_id)
+    # get all pages of the same tree
+    pages = region.pages.filter(archived=False, tree_id=page.tree_id)
+    page_translation = page.get_translation(language_code)
+    language = Language.objects.get(code=language_code)
+    template_path = "pages/page_view.html"
+    context = {
+        "page_translation": page_translation,
+        "html2pdf": True,
+        "region": region,
+        "page_level": page.level,
+        "pages": pages,
+        "language": language,
+    }
+    response = HttpResponse(content_type="application/pdf")
+    response["Content-Disposition"] = (
+        "attachment; filename=%s.pdf" % page_translation.title
+    )
```

Comment:
Since the created PDF is offered for download to the user, i suggest to use the local download folder as the cache. So i could just check this folder whether that hash already exists.

Reply:
I'm not sure if I understood this correctly. Do you mean the local download folder of the user? The "expensive" operation is not the actual download of the file, but the server-side creation of the PDF. So what we want to achieve with the cache is that we store created PDFs on the server, so that when another user wants to download the same PDF, `pisa.CreatePDF` is not called again, but the PDF file from the cache is used. We don't need to take care about when the same user downloads the same file twice.

Reply:
All right, so I will try to figure out, which is the recommended place for such cache inside our project folder.

Reply:
You could e.g. use `src/cms/static/CACHE`, which is also used by the django-compressor.